### PR TITLE
Clean up BlockSnapshot code

### PR DESCRIPTION
**Pull request by bs2609**
This cleans up the code for `BlockSnapshot`, and also fixes a few minor bugs for good measure.

Changes so far:
*  Replaced mostly copy-pasted code with calls to existing methods.
 This also fixes a location where `BlockPos.toImmutable()` was not being called.
*  Cleaned up `equals()` and `hashCode()`
  Reordered comparisons to ensure faster checks are first and used Guava's null-safe `equals()` for brevity.
  Also removed world field from these methods - we have the dimension ID for that.
*  Fixed logic in `readFromNBT()` where "hasTE" being true set the NBT tag to null, and a false value set it to an empty compound tag.
*  The world is now stored via a `WeakReference` to avoid leaking it.

Other things I'd like opinions on:
*  `DimensionManager.getWorld()` is currently used. This can return null if the dimension in question is not loaded - does this constitute a problem? Should `MinecraftServer.worldServerForDimension()`, which loads dimensions if needed, be used instead?
  (`DimensionManager.getWorld()` should probably be annotated `@Nullable`, but that's for another time)
*  The code produces some debug output through `System.out.printf()` calls, should these be run though the FML log?
*  I'm confused by some of the current behavior for restoring snapshots.
    *  If the block state does not appear to be the same across the snapshot, we call `world.setBlockState()` if 'force' is true, otherwise we just return false.
After this, we call `world.setBlockState()`, possibly for the second time, and then `world.notifyBlockUpdate()`.
    * The original version of the code (1.7 era, see link) had the first `setBlockState()` call (inside the check) as `world.setBlock()`, and the second call as `world.setBlockMetadataWithNotify()`. I'm not too familiar with 1.7, is it safe to assume only one call would be needed now?
    * Do we need to call `setBlockState()` at all if 'force' is false and we don't return early?
    * Is the call to `notifyBlockUpdate()` needed, as the flags passed to `setBlockState` always contain '2' and so it'll be called from `markAndNotifyBlock()` anyway, which appears like it will be called by `setBlockState()` as a matter of course.

Original version of code [here](https://github.com/MinecraftForge/MinecraftForge/blob/37f91681d97fd388b18cdf5b60b139b4d69792f5/src/main/java/net/minecraftforge/common/util/BlockSnapshot.java)

Longer term issues for later:
*  We should probably look into changing the format used for serialisation, using `NBTUtil` to store the `BlockPos` and `IBlockState` fields in compound tags. In particular the current (string pair plus int meta) format for blockstates seems very not-future-proof. 

(Sorry about the wall of text)

### CODE REVIEW

File: src/main/java/net/minecraftforge/common/util/BlockSnapshot.java [hunk 1]
```diff
@@ -135,18 +120,28 @@ public static BlockSnapshot readFromNBT(NBTTagCompound tag)
...
+
     public IBlockState getCurrentBlock()
     {
         return getWorld().getBlockState(getPos());
     }
 
     public World getWorld()
     {
-        if (this.world == null)
+        World world = this.world != null ? this.world.get() : null;
```
Comment (mezz) [inline src/main/java/net/minecraftforge/common/util/BlockSnapshot.java:143]: I think you can avoid the ternary statement here by checking `if (this.world == null)` instead, and returning `this.world.get()` at the end.

File: src/main/java/net/minecraftforge/common/util/BlockSnapshot.java [hunk 2]
```diff
@@ -50,30 +52,13 @@
     private int flag;
     @Nullable
     private final NBTTagCompound nbt;
-    private transient World world;
+    private transient WeakReference<World> world;
```
Comment (mezz) [inline src/main/java/net/minecraftforge/common/util/BlockSnapshot.java]: You could mark this as `@Nullable` for good measure

File: src/main/java/net/minecraftforge/common/util/BlockSnapshot.java (hunk 1 above)
Comment (bs2609) [reply src/main/java/net/minecraftforge/common/util/BlockSnapshot.java:143]: It'd have to be `if (this.world == null || this.world.get() == null)` in that case, the ternary isn't great but we don't have `?.` as an operator in Java.

File: src/main/java/net/minecraftforge/common/util/BlockSnapshot.java (hunk 2 above)
Comment (bs2609) [reply src/main/java/net/minecraftforge/common/util/BlockSnapshot.java]: Done

File: src/main/java/net/minecraftforge/common/util/BlockSnapshot.java (hunk 1 above)
Comment (mezz) [reply src/main/java/net/minecraftforge/common/util/BlockSnapshot.java:143]: Oh I get it now, the weak reference being held will become null if the world unloads.

File: src/main/java/net/minecraftforge/common/util/BlockSnapshot.java [hunk 3]
```diff
@@ -309,25 +273,30 @@ public boolean equals(Object obj)
     public int hashCode()
     {
         int hash = 7;
-        hash = 73 * hash + this.getPos().getX();
-        hash = 73 * hash + this.getPos().getY();
-        hash = 73 * hash + this.getPos().getZ();
         hash = 73 * hash + this.getMeta();
         hash = 73 * hash + this.getDimId();
+        hash = 73 * hash + this.getPos().hashCode();
```
Comment (liach) [inline src/main/java/net/minecraftforge/common/util/BlockSnapshot.java]: If you change hashcode, how about backward compatibility?

Comment (diesieben07) [reply src/main/java/net/minecraftforge/common/util/BlockSnapshot.java]: hashCode is not required to be consistent even between JVM runs. Relying on it is a bug.

**Comment by mezz** [comment]
I've never looked at `BlockSnapshot` before so it may take me a while to review.
Can you briefly explain its purpose so I can get started easier when I go to review it?

**Comment by bs2609** [comment]
The original pull request (#1367) might help explain a bit. The main usage in Forge is `ForgeHooks.onPlaceItemIntoWorld()`.

**Comment by mezz** [comment]
I'll try to give opinions on these questions as I understand more of this system.

>DimensionManager.getWorld() is currently used. This can return null if the dimension in question is not loaded - does this constitute a problem? Should MinecraftServer.worldServerForDimension(), which loads dimensions if needed, be used instead?

I don't think you can make a block snapshot in an unloaded world. Is that possible?
It seems that they are created, used in events, and then used to restore if the event says to.

In fact I don't see any uses of `BlockSnapshot#readFromNBT` or `BlockSnapshot#writeToNBT` or a reason for extending `Serializable` which really confused me when I was trying to figure out the purpose of `BlockSnapshot`.

I assume that `BlockSnapshot` was originally written for another project, I can't think of another good reason for the unused functionality unless I'm still misunderstanding how it gets used. @bloodmc can you confirm?

I think any unused methods should be deprecated and eventually removed to help with clarity and maintenance. Can you also remove the `@SuppressWarnings({"serial", "deprecation"})` at the top?

>The code produces some debug output through System.out.printf() calls, should these be run though the FML log?

These are probably never used, or almost never. I think it's worth leaving it this way so it stands out when people do have to debug it.

**Comment by bs2609** [comment]
I've removed references to `Serializable` from the class - while this is _technically_ a breaking change, as mentioned [here](http://www.minecraftforge.net/forum/topic/55008-1112-blocksnapshot-serialization-not-posible/), `BlockSnapshot` can't actually be serialised at present.

**Comment by mezz** [comment]
@bs2609 have you made sense of this? You had a lot of requests for opinions before and I'm not sure where this stands now.

**Comment by bs2609** [comment]
I'd mostly like advice on what the 1.7.10 behaviour for restoring snapshots looked like.

**Comment by mezz** [comment]
Try to get in touch with blood on IRC in `#spongedev`, I'm not sure anyone else knows.

**Comment by liach** [comment]
I suggest to go on [Sponge discord channel](https://discord.gg/P32HEwh) or [GriefPrevention discord channel](https://discord.gg/mGC9D4D) to find him. He is always on GP.

**Comment by mezz** [comment]
@bs2609 Any luck contacting blood?

**Comment by mezz** [comment]
@bs2609 if you still haven't been able to get in contact, I think you should do whatever you think makes sense.
Also, this PR should be updated to 1.12.

**Comment by mezz** [comment]
I am closing this because there has been no response for a long time.
Feel free to reopen it if you make some progress.

**Comment by bloodmc** [comment]
This PR looks good to me. The only issue is the removal of Serializable. This was added so mods could serialize BlockSnapshot objects for things like schematics. IBlockState and World would have to be transient if kept which essentially ignores serialization of those objects. Attempting to serialize World would be very very bad =). 

Update: I see there was an issue with BlockPos being serialized. BlockSnapshot was added before BlockPos even existed. To workaround this issue, you can simply store BlockPos as ints.

**Comment by bs2609** [comment]
Is there any reason why NBT serialisation is not adequate for those purposes? I'd rather use the code which already exists to handle this, rather than needing to come up with another custom way to serialise blockstates in future versions.

(also see this [gist](https://gist.github.com/bs2609/eafd5322423d3f03a02ada2bf5779a48))

**Comment by bs2609** [comment]
Updated to target 1.12.x.

**Comment by bs2609** [comment]
@mezz, mind relabeling this?

**Comment by mezz** [comment]
Talked to blood and they said nbt serialization is probably just fine.

**Comment by bs2609** [comment]
If the serialisation changes are fine, then I think this is 'done' now.

**Pull request merged**